Immediate panic when HELMFILE_DISABLE_INSECURE_FEATURES environment variable is set to true

### Operating system

Windows, Fedora

### Helmfile Version

v1.1.1, v0.168.0, probably more

### Helm Version

3.15.2, 3.16.1, but does not seem to be applicable in this case

### Bug description

When the environment variable `HELMFILE_DISABLE_INSECURE_FEATURES` is set to `true`, helmfile panics and exits with a stack trace.

I have reproduced this issue in multiple scenarios, including inside an empty directory. It appears that almost every command causes the same panic to occur.

Unfortunately, I am unable to provide a stack trace due to... work :eyes:. Instead, here is a set of steps that should produce the same results as what I am describing:

1. In a fresh terminal, create a new directory and enter it
2. Set `HELMFILE_DISABLE_INSECURE_FEATURES` to `true`
3. Run something like `helmfile write-values`

First line of output should read:

```
panic: Remote sources are disabled due to ‘DISABLE_INSECURE_FEATURES’
```

...followed by the stack trace.

### Example helmfile.yaml

No helmfile needed, but even the most basic one will do.

### Error message you've seen (if any)

panic: Remote sources are disabled due to ‘DISABLE_INSECURE_FEATURES’

(followed by a stack trace that I am regrettably unable to provide)

### Steps to reproduce

see issue description

### Working Helmfile Version

currently unknown

### Relevant discussion

https://github.com/helmfile/helmfile/discussions/2069

Following @JimmyAllnighter steps to reproduce induced suggested behaviour first go.

image

```
registry.access.redhat.com/ubi8/ubi
```


bin dir
```
[root@ceefeae5c73a helmbin]# ls -larth
total 250M
-rwxr-xr-x 1 1001  127  51M Jun 12  2024 helm-3.15.2
-rwxr-xr-x 1 1001  118  57M Jun  2 13:00 helm-3.18.2
drwxr-xr-x 1 root root 4.0K Jun  6 11:57 ..
-rw-r--r-- 1 root root   31 Jun  6 11:58 helmfile.yaml
-rwxr-xr-x 1 root root 110M Jun  6 11:58 helmfile
-rwxr-xr-x 1 root root  33M Jun  6 11:58 helmfile_1.1.1_linux_amd64.tar.gz
drwxr-xr-x 2 root root 4.0K Jun  6 11:58 .
```

``` 
[root@ceefeae5c73a helmbin]# pwd
/helmbin
```

---
helmfile.yaml (helm 3.15.2)

```
helmBinary: ./helm-3.15.2
```

---
printenv
```
[root@ceefeae5c73a helmbin]# printenv
LANG=C.utf8
HOSTNAME=ceefeae5c73a
OLDPWD=/var/bins
which_declare=declare -f
container=oci
PWD=/helmbin
HOME=/root
TERM=xterm
SHLVL=1
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HELMFILE_DISABLE_INSECURE_FEATURES=true
BASH_FUNC_which%%=() {  ( alias;
 eval ${which_declare} ) | /usr/bin/which --tty-only --read-alias --read-functions --show-tilde --show-dot $@
}
_=/usr/bin/printenv
```

---

test dir

```
[root@ceefeae5c73a helmtest]# ls
[root@ceefeae5c73a helmtest]# pwd
/helmtest
```

---

**demonstrated error - helm 3.15.2**

```
[root@ceefeae5c73a helmtest]# /helmbin/helmfile -f /helmbin/helmfile.yaml write-values
panic: Remote sources are disabled due to 'DISABLE_INSECURE_FEATURES'

goroutine 66 [running]:
github.com/helmfile/helmfile/pkg/remote.NewRemote(0x4192d4?, {0x0?, 0xc000b1eb58?}, 0x419d65?)
        /home/runner/work/helmfile/helmfile/pkg/remote/remote.go:536 +0x1d2
github.com/helmfile/helmfile/pkg/app.(*App).visitStatesWithSelectorsAndRemoteSupport(0xc000b466c0, {0x7ffd33314e35, 0x16}, 0xc000b1ec48, 0x0, {0xc000b1ecb0, 0x1, 0x3daf901?})
        /home/runner/work/helmfile/helmfile/pkg/app/app.go:1096 +0x2fe
github.com/helmfile/helmfile/pkg/app.(*App).ForEachState(0xc000b466c0, 0xc000b1ecb8, 0x0, {0xc000b1ecb0, 0x1, 0x1})
        /home/runner/work/helmfile/helmfile/pkg/app/app.go:961 +0xab
github.com/helmfile/helmfile/pkg/app.(*App).WriteValues(0xc000b466c0, {0x44fa718, 0xc000a048c0})
        /home/runner/work/helmfile/helmfile/pkg/app/app.go:252 +0x93
github.com/helmfile/helmfile/cmd.NewWriteValuesCmd.func1(0xc00050ef00?, {0x3daa5e8?, 0x4?, 0x3daa6d0?})
        /home/runner/work/helmfile/helmfile/cmd/write-values.go:29 +0xfb
github.com/spf13/cobra.(*Command).execute(0xc000b54608, {0xc000b28600, 0x2, 0x2})
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1015 +0xa94
github.com/spf13/cobra.(*Command).ExecuteC(0xc0004ae608)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1148 +0x40c
github.com/spf13/cobra.(*Command).Execute(0xc000b18000?)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1071 +0x13
main.main.func1()
        /home/runner/work/helmfile/helmfile/main.go:29 +0x69
created by main.main in goroutine 1
        /home/runner/work/helmfile/helmfile/main.go:21 +0xcf

```

---
helmfile.yaml (helm 3.18.2)

```
helmBinary: ./helm-3.18.2
```

---

**demonstrated error - helm 3.18.2**

```
[root@ceefeae5c73a helmtest]# /helmbin/helmfile -f /helmbin/helmfile.yaml write-values
panic: Remote sources are disabled due to 'DISABLE_INSECURE_FEATURES'

goroutine 24 [running]:
github.com/helmfile/helmfile/pkg/remote.NewRemote(0x4192d4?, {0x0?, 0xc00093cb58?}, 0x419d65?)
        /home/runner/work/helmfile/helmfile/pkg/remote/remote.go:536 +0x1d2
github.com/helmfile/helmfile/pkg/app.(*App).visitStatesWithSelectorsAndRemoteSupport(0xc0007c1200, {0x7ffcc2e74e35, 0x16}, 0xc00093cc48, 0x0, {0xc00093ccb0, 0x1, 0x3daf901?})
        /home/runner/work/helmfile/helmfile/pkg/app/app.go:1096 +0x2fe
github.com/helmfile/helmfile/pkg/app.(*App).ForEachState(0xc0007c1200, 0xc00093ccb8, 0x0, {0xc00093ccb0, 0x1, 0x1})
        /home/runner/work/helmfile/helmfile/pkg/app/app.go:961 +0xab
github.com/helmfile/helmfile/pkg/app.(*App).WriteValues(0xc0007c1200, {0x44fa718, 0xc0005de070})
        /home/runner/work/helmfile/helmfile/pkg/app/app.go:252 +0x93
github.com/helmfile/helmfile/cmd.NewWriteValuesCmd.func1(0xc0001da400?, {0x3daa5e8?, 0x4?, 0x3daa6d0?})
        /home/runner/work/helmfile/helmfile/cmd/write-values.go:29 +0xfb
github.com/spf13/cobra.(*Command).execute(0xc000a13b08, {0xc00095c5a0, 0x2, 0x2})
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1015 +0xa94
github.com/spf13/cobra.(*Command).ExecuteC(0xc00057bb08)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1148 +0x40c
github.com/spf13/cobra.(*Command).Execute(0xc0008a4640?)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1071 +0x13
main.main.func1()
        /home/runner/work/helmfile/helmfile/main.go:29 +0x69
created by main.main in goroutine 1
        /home/runner/work/helmfile/helmfile/main.go:21 +0xcf

```

--- a/pkg/remote/remote.go
+++ b/pkg/remote/remote.go
@@ -190,6 +190,11 @@ func (r *Remote) Fetch(path string, cacheDirOpt ...string) (string, error) {
 		return "", err
 	}
 
+	// Block remote access if insecure features are disabled and the source is remote
+	if disableInsecureFeatures && IsRemote(path) {
+		return "", fmt.Errorf("remote sources are disabled due to 'HELMFILE_DISABLE_INSECURE_FEATURES'")
+	}
+
 	srcDir := fmt.Sprintf("%s://%s/%s", u.Scheme, u.Host, u.Dir)
 	file := u.File
 
@@ -532,9 +537,6 @@ func ParseS3Url(s3URL string) (string, string, error) {
 }
 
 func NewRemote(logger *zap.SugaredLogger, homeDir string, fs *filesystem.FileSystem) *Remote {
-	if disableInsecureFeatures {
-		panic("Remote sources are disabled due to 'DISABLE_INSECURE_FEATURES'")
-	}
 	remote := &Remote{
 		Logger:     logger,
 		Home:       homeDir,

